src/Microsoft.TemplateEngine.TemplateLocalizer.Core/TraversalRules/RegexFilteredTraversalRule.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.RegularExpressions;
+using Microsoft.TemplateEngine.TemplateLocalizer.Core.KeyExtractors;
+
+namespace Microsoft.TemplateEngine.TemplateLocalizer.Core.TraversalRules
+{
+    /// <summary>
+    /// Filters identifiers based on whether they match with a given regex.
+    /// </summary>
+    internal sealed class RegexFilteredTraversalRule : TraversalRule
+    {
+        private readonly Regex _regex;
+
+        /// <summary>
+        /// Creates an instance of <see cref="RegexFilteredTraversalRule"/>.
+        /// </summary>
+        /// <param name="regexPattern">Regex pattern string that will be used to match json element names.</param>
+        /// <param name="keyCreator"><see cref="IJsonKeyCreator"/> to be used when creating a key for the match elements.</param>
+        public RegexFilteredTraversalRule(string regexPattern, IJsonKeyCreator? keyExtractor = null)
+            : this(new Regex(regexPattern), keyExtractor)
+        { }
+
+        /// <summary>
+        /// Creates an instance of <see cref="RegexFilteredTraversalRule"/>.
+        /// </summary>
+        /// <param name="regex">Regex pattern that will be used to match json element names.</param>
+        /// <param name="keyCreator"><see cref="IJsonKeyCreator"/> to be used when creating a key for the match elements.</param>
+        public RegexFilteredTraversalRule(Regex regex, IJsonKeyCreator? keyExtractor = null)
+            : base(keyExtractor)
+        {
+            _regex = regex;
```

Comment:
should we validate the Regex is not null?

Reply:
Regex here is marked as "non-nullable", so passing something potentially null will generate a compile time error. This type is also internal. So check shouldn't be necessary.

Reply:
Validations should not be driven by the access modifier, but by the impact bad state would have.